File: .idea/CombinationSum.java
class CombinationSum {
    List<List<Integer>> result;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList();
        if(candidates.length==0 || candidates == null)
            return result;
        helper(candidates, target, 0, new ArrayList<>());
        return result;

    }
    private void helper(int[] candidates, int target, int index, List<Integer> path)
    {
        if(target ==0)
        {
            result.add(new ArrayList<>(path));
            return;
        }
        if(target < 0)
            return;
        for(int i = index; i< candidates.length;i++)
        {
            path.add(candidates[i]);
            helper(candidates, target - candidates[i], i, path);

            path.remove(path.size()-1);
        }
    }
}

File: CombinationSUM.py
import copy
class Solution:
    result = []

    def combinationSum(self, candidates, target) :

        self.result = []

        if candidates == None or len(candidates) == 0:
            return self.result
        self.helper(candidates, target, 0, [])
        return self.result

    def helper(self, candidates, target, index, path):

        if target == 0:
            self.result.append(copy.deepcopy(path))
            return
        if target < 0 or index == len(candidates):
            return

        # dont choose
        # curr_path = []
        # curr_path.extend(path)
        self.helper(candidates, target, index + 1, path)

        # choose
        # curr_path1 = []
        # curr_path1.extend(path)
        path.append(candidates[index])
        self.helper(candidates, target - candidates[index], index, path)
        path.pop(-1)